From a8e1567bfb132e27312b584a13e37e3bb20124af Mon Sep 17 00:00:00 2001
From: Chen Jiahao <chenjiahao16@huawei.com>
Date: Mon, 18 Apr 2022 16:58:13 +0800
Subject: [PATCH 2676/4585] [Huawei] ARM: spectre-bhb: turn off the mitigation
 via boot cmdline param

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186682

--------------------------------

HULK 5.10 enables spectre-bhb mitigation by default for ARM32, which may
cause performance regression. To offer products an option to turn off
the mitigation, implement a cmdline parameter 'nospectre_bhb' compatible
with mainline, which sets up a switch to skip the mitigation code entry
and bhb workarounds in mitigation method handler.

Signed-off-by: Chen Jiahao <chenjiahao16@huawei.com>
---
 arch/arm/mm/proc-v7-bugs.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/arch/arm/mm/proc-v7-bugs.c b/arch/arm/mm/proc-v7-bugs.c
index 6ecefc6962e5..c91dc11e65cd 100644
--- a/arch/arm/mm/proc-v7-bugs.c
+++ b/arch/arm/mm/proc-v7-bugs.c
@@ -15,6 +15,8 @@
  */
 int nospectre_v2 __read_mostly;
 
+static int nospectre_bhb __read_mostly;
+
 static int __init nospectre_v2_setup(char *str)
 {
 	nospectre_v2 = 1;
@@ -22,6 +24,13 @@ static int __init nospectre_v2_setup(char *str)
 }
 early_param("nospectre_v2", nospectre_v2_setup);
 
+static int __init nospectre_bhb_setup(char *str)
+{
+	nospectre_bhb = 1;
+	return 0;
+}
+early_param("nospectre_bhb", nospectre_bhb_setup);
+
 #ifdef CONFIG_ARM_PSCI
 static int __maybe_unused spectre_v2_get_cpu_fw_mitigation_state(void)
 {
@@ -245,6 +254,11 @@ static void cpu_v7_spectre_bhb_init(void)
 {
 	unsigned int state, method = 0;
 
+	if (nospectre_bhb) {
+		pr_info_once("Spectre bhb: hardening is disabled\n");
+		return;
+	}
+
 	switch (read_cpuid_part()) {
 	case ARM_CPU_PART_CORTEX_A15:
 	case ARM_CPU_PART_BRAHMA_B15:
-- 
2.17.1

